File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"fmt"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const BIRTHDATE_LEN = 10
const UINT32_SIZE = 4
const UINT8_SIZE = 1
const FRAME_SIZE = 5

type Opcode byte

const (
	OpBatch Opcode = iota
	OpEnd
	OpACK
	OpWinners
	OpHello
)

type ClientProtocol struct {
	socket   safe_socket.Socket
	agencyId int
}

type Frame struct {
	opcode Opcode
	length int
}

func CrearProtocolo(socket safe_socket.Socket, agencyId int) *ClientProtocol {
	return &ClientProtocol{socket: socket, agencyId: agencyId}
}

func (p *ClientProtocol) SendBatch(bets []lottery.Bet) error {
	var payload []byte 
	// agregamos al principio del batch la cantidad de bets dentro del batch
	payload = binary.BigEndian.AppendUint32(payload, uint32(len(bets)))

	for _, bet := range bets {
		serialized_bet, err := p.serializeBet(bet)
		if err != nil {
			return err
		}
		payload = append(payload, serialized_bet...)
	}

	return p.sendFramed(payload, OpBatch)
}

func (p *ClientProtocol) SendHello() error {
	var payload []byte

	payload = append(payload, byte(p.agencyId))
	return p.sendFramed(payload, OpHello)
}

func (p *ClientProtocol) sendFramed(payload []byte, opcode Opcode) error {
	var frame []byte

	// primero el flag de fin
	frame = append(frame, byte(opcode))
	// luego el largo
	frame = binary.BigEndian.AppendUint32(frame, uint32(len(payload)))
	// finalmente el payload
	frame = append(frame, payload...)

	err := p.socket.Send(frame)
	if err != nil {
		return err
	}
	return nil
}

func (p *ClientProtocol) serializeBet(bet lottery.Bet) ([]byte, error) {
	var buf []byte

	// se agrega al buffer 1 byte de longitud + string para nombre
	buf = append(buf, byte(len(bet.FirstName)))
	buf = append(buf, []byte(bet.FirstName)...)

	// se agrega al buffer 1 byte de longitud + string para apellido
	buf = append(buf, byte(len(bet.LastName)))
	buf = append(buf, []byte(bet.LastName)...)

	// se agrega al buffer BirthdateLen para la fecha (tamaño fijo para fechas)
	buf = append(buf, []byte(bet.Birthdate)...)

	buf = binary.BigEndian.AppendUint32(buf, uint32(bet.Document))
	buf = binary.BigEndian.AppendUint32(buf, uint32(bet.Number))

	return buf, nil
}

func (p *ClientProtocol) SendEnd() error{
	var buf []byte
	return p.sendFramed(buf, OpEnd)
}

func (p *ClientProtocol) Close() error {
	return p.socket.Close()
}

func (p *ClientProtocol) RecvWinners() ([]lottery.Bet, error) {
	frame, err := p.readFrame()
	if err != nil {
		return nil, err
	}

	payload, err := p.socket.Recv(frame.length)
	if err != nil {
		return nil, err
	}

	count := binary.BigEndian.Uint32(payload[:UINT32_SIZE])
	payload = payload[UINT32_SIZE:]

	winners := make([]lottery.Bet, 0, count)	
	for i := uint32(0); i < count; i++ {
		var bet lottery.Bet
		bet, payload = p.deserializeBet(payload)
		winners = append(winners, bet)
	}

	return winners, nil
}

func (p *ClientProtocol) readFrame() (Frame, error) {
	frame_bytes, err := p.socket.Recv(FRAME_SIZE)
	if err != nil {
		return Frame{}, err
	}
	opcode := Opcode(frame_bytes[0])
	frame_bytes = frame_bytes[UINT8_SIZE:]
	length := int(binary.BigEndian.Uint32(frame_bytes[:UINT32_SIZE]))

	return Frame{opcode, length}, nil
}

func (p *ClientProtocol) deserializeBet(payload []byte) (lottery.Bet, []byte) {
	fnLen := int(payload[0])
	payload = payload[UINT8_SIZE:]
	firstName := string(payload[:fnLen])
	payload = payload[fnLen:]

	lnLen := int(payload[0])
	payload = payload[UINT8_SIZE:]
	lastName := string(payload[:lnLen])
	payload = payload[lnLen:]

	birthdate := string(payload[:BIRTHDATE_LEN])
	payload = payload[BIRTHDATE_LEN:]

	document := int(binary.BigEndian.Uint32(payload[:UINT32_SIZE]))
	payload = payload[UINT32_SIZE:]

	number := int(binary.BigEndian.Uint32(payload[:UINT32_SIZE]))
	payload = payload[UINT32_SIZE:]

	bet := lottery.Bet{
		FirstName: firstName,
		LastName:  lastName,
		Document:  document,
		Birthdate: birthdate,
		Number:    number,
	}
	return bet, payload
}

func (p *ClientProtocol) RecvACK() (Opcode, error){
	frame, err := p.readFrame()
	if err != nil {
		return OpEnd, err
	}
	if frame.opcode != OpACK { return OpEnd, fmt.Errorf("ack inesperado: se recibió opcode %d", frame.opcode) }
	return frame.opcode, nil
}

File: services/server/src/protocol/protocol.py
from dataclasses import dataclass

import safe_socket as ss
from lottery import Bet


FRAME_LEGNTH = 5
BIRTHDATE_LEN = 10
UINT8_SIZE = 1
UINT32_SIZE = 4

@dataclass
class Frame:
    agency_id: int
    length: int

class Protocol:
    def __init__(self, socket):
        self.socket = socket

    def recv_batch(self) -> tuple[list[Bet], int]:
        frame = self._read_frame()
        batch = []

        if self._is_last_batch(frame):
            return batch, frame.agency_id

        payload = ss.recv_all(self.socket, frame.length)

        batch_size = int.from_bytes(payload[:UINT32_SIZE], "big")
        payload = payload[UINT32_SIZE:]
        
        for _ in range (batch_size):
            bet, payload = self._deserialize_bet(payload, frame)
            batch.append(bet)

        return batch, frame.agency_id

    def _read_frame(self) -> Frame:
        frame_bytes = ss.recv_all(self.socket, FRAME_LEGNTH)

        agency_id = int.from_bytes(frame_bytes[:UINT8_SIZE], "big")
        length = int.from_bytes(frame_bytes[UINT8_SIZE:], "big")

        return Frame (
            agency_id,
            length,
        )

    def _deserialize_bet(self, payload, frame) -> Bet:
        fn_length = payload[0]
        payload = payload[UINT8_SIZE:]

        first_name =  payload[:fn_length].decode()
        payload = payload[fn_length:]        

        ln_length = int.from_bytes(payload[:UINT8_SIZE], "big")
        payload = payload[UINT8_SIZE:]

        last_name =  payload[:ln_length].decode()
        payload = payload[ln_length:]

        birthday = payload[:BIRTHDATE_LEN].decode()
        payload = payload[BIRTHDATE_LEN:]

        document = int.from_bytes(payload[:UINT32_SIZE], "big")
        payload = payload[UINT32_SIZE:]

        number = int.from_bytes(payload[:UINT32_SIZE], "big")
        payload = payload[UINT32_SIZE:]

        return Bet(
            frame.agency_id,
            first_name,
            last_name,
            document,
            birthday,
            number
        ), payload

    def send_winners(self, winners: list[Bet]) -> None:
        payload = b""

        # agregamos cantidad de ganadores al payload
        payload += len(winners).to_bytes(UINT32_SIZE, "big")
        for winner in winners:
            serialized = self._serialize_bet(winner)
            payload += serialized

        self._send_framed(payload)

    def _send_framed(self, payload):
        packet = b""

        # agregamos longitud total del paquete y payload
        packet += len(payload).to_bytes(UINT32_SIZE, "big")
        packet += payload

        ss.send_all(self.socket, packet)

    def _serialize_bet(self, bet : Bet) -> bytes:
        b_bet = b""

        fm_bytes = bet.first_name.encode()
        # 1 byte para longitud del nombre
        b_bet += len(fm_bytes).to_bytes(UINT8_SIZE, "big")
        b_bet += fm_bytes

        lm_bytes = bet.last_name.encode()
        # 1 byte para longitud del apellido
        b_bet += len(lm_bytes).to_bytes(UINT8_SIZE, "big")
        b_bet += lm_bytes

        b_bet += bet.birthdate.encode()

        b_bet += bet.document.to_bytes(UINT32_SIZE, "big")
        b_bet += bet.number.to_bytes(UINT32_SIZE, "big")

        return b_bet


    def _is_last_batch(self, frame) -> bool:
        return frame.length == 0